[2/n] Thread PG: add class _World to distributed_c10d.py (#781)

yhcharles opened the pull request:
Summary:
X-link: https://github.com/pytorch/torchrec/pull/781

Move a bunch of globals to instance methods and replace all use to them.

We move all PG related globals under World and use a singleton instance under _world.

This creates an undocumented extension point to inject full control of how how c10d
state behaves.

One simple hack is to change _world to an implementation that uses a threadlocal
and enable per-thread PGs.

It almost get DDP working and the PG is missing an implementation of all_reduce.

This enables notebook usage of PTD, which is a big deal for learning it:
https://gist.github.com/kumpera/32cb051fa26b8cad8bdf671f968dcd68

This change ensures BC by keeping the global variables around and have the default _World wrap it.

I have relinked this diff to a new github PR, so that I can update it. The original PR is
> Pull Request resolved: https://github.com/pytorch/pytorch/pull/86348

Differential Revision: D40236769

Pulled By: yhcharles

gnadathur reviewed (approved):
LGTM.

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -273,32 +273,113 @@ def __getattribute__(self, key):
 
-class group(object):
+# DO NOT USE THESE FIELDS DIRECTLY.
+# Use them through the _world object to make sure the _world override mecanism
```
nit: typo, "mechanism"

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -273,32 +273,113 @@ def __getattribute__(self, key):
-class group(object):
+# DO NOT USE THESE FIELDS DIRECTLY.
+# Use them through the _world object to make sure the _world override mecanism
+_pg_map: Dict[ProcessGroup, Tuple[str, Optional[Store]]] = {}
+_pg_names: Dict[ProcessGroup, str] = {}
+_pg_group_ranks: Dict[ProcessGroup, Dict[int, int]] = {}
+_group_count = 0
```
Are these fields only used by `class _World`? Why not put them under that class

yhcharles replied:
They are used in some other projects. So we have to keep them. Let me ping you some examples.

wanchaol replied:
Hey I am randomly browsed into this piece of code. I am wondering why we need to keep them? I know it's being used in other projects, but I feel since we make all those variables be private variables, shouldn't that mean we are free to change this to whatever we like without worrying about BC?

If it's internal change related, maybe we can just refactor it directly.

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -883,8 +962,8 @@ def _new_process_group_helper(
                         world_size=group_size,
                         timeout=timeout,
                     )
-            _pg_map[pg] = (Backend.GLOO, store)
```
FYI I have a PR to refactor some of this code since this and the if statements below are basically repeated, https://github.com/pytorch/pytorch/pull/88351. Feel free to land yours first if you are ready though

rohan-varma reviewed (approved):
Stamp to unblock, but please address @H-Huang's comments

kumpera committed:
[2/n] Thread PG: add class _World to distributed_c10d.py (#88471)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88471

X-link: https://github.com/pytorch/torchrec/pull/781

Move a bunch of globals to instance methods and replace all use to them.

We move all PG related globals under World and use a singleton instance under _world.

This creates an undocumented extension point to inject full control of how how c10d
state behaves.

One simple hack is to change _world to an implementation that uses a threadlocal
and enable per-thread PGs.

It almost get DDP working and the PG is missing an implementation of all_reduce.

This enables notebook usage of PTD, which is a big deal for learning it:
https://gist.github.com/kumpera/32cb051fa26b8cad8bdf671f968dcd68

This change ensures BC by keeping the global variables around and have the default _World wrap it.

I have relinked this diff to a new github PR, so that I can update it. The original PR is
> Pull Request resolved: https://github.com/pytorch/pytorch/pull/86348

Reviewed By: gnadathur

Differential Revision: D40236769

Pulled By: yhcharles

fbshipit-source-id: efeaa7990e26a58987769a93cedf7318d5cae445
